sound/soc/sof/amd/acp-pcm.c

```diff
@@ -83,3 +84,88 @@ int acp_pcm_close(struct snd_sof_dev *sdev, struct snd_pcm_substream *substream)
 	return acp_dsp_stream_put(sdev, stream);
 }
 EXPORT_SYMBOL_NS(acp_pcm_close, SND_SOC_SOF_AMD_COMMON);
+
+static const struct acp_pcm_table pcm_dev[] = {
+	{I2S_BT, "I2SBT"},
+	{I2S_SP, "I2SSP"},
+	{PDM_DMIC, "DMIC"},
+};
+
+static enum acp_pcm_types get_id_from_list(const char *name)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(pcm_dev); i++) {
+		if (!strcmp(name, pcm_dev[i].pcm_name))
+			return pcm_dev[i].pcm_index;
+	}
+
+	return PCM_NONE;
+}
+
+static u64 acp_get_byte_count(struct snd_sof_dev *sdev, struct snd_pcm_substream *substream,
+				enum acp_pcm_types pcm_id)
+{
+	u64 bytescount, low = 0, high = 0;
+	u32 reg1 = 0, reg2 = 0;
+
+	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
+		switch (pcm_id) {
+		case I2S_BT:
+			reg1 = ACP_BT_TX_LINEARPOSITIONCNTR_HIGH;
+			reg2 = ACP_BT_TX_LINEARPOSITIONCNTR_LOW;
+			break;
+		case I2S_SP:
+			reg1 = ACP_I2S_TX_LINEARPOSITIONCNTR_HIGH;
+			reg2 = ACP_I2S_TX_LINEARPOSITIONCNTR_LOW;
+			break;
+		default:
+			return -EINVAL;
+		}
+	} else {
+		switch (pcm_id) {
+		case I2S_BT:
+			reg1 = ACP_BT_RX_LINEARPOSITIONCNTR_HIGH;
+			reg2 = ACP_BT_RX_LINEARPOSITIONCNTR_LOW;
+			break;
+		case I2S_SP:
+			reg1 = ACP_I2S_RX_LINEARPOSITIONCNTR_HIGH;
+			reg2 = ACP_I2S_RX_LINEARPOSITIONCNTR_LOW;
+			break;
+		case PDM_DMIC:
+			reg1 = ACP_WOV_RX_LINEARPOSITIONCNTR_HIGH;
+			reg2 = ACP_WOV_RX_LINEARPOSITIONCNTR_LOW;
+			break;
+		default:
+			return -EINVAL;
+		}
+	}
+
+	high = snd_sof_dsp_read(sdev, ACP_DSP_BAR, reg1);
+	low = snd_sof_dsp_read(sdev, ACP_DSP_BAR, reg2);
+
+	bytescount = (high << 32) | low;
+	return bytescount;
+}
+
+snd_pcm_uframes_t acp_pcm_pointer(struct snd_sof_dev *sdev, struct snd_pcm_substream *substream)
+{
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct snd_soc_component *scomp = sdev->component;
+	enum acp_pcm_types pcm_id = PCM_NONE;
+	struct snd_sof_pcm *spcm;
+	u32 pos, buffersize;
+	u64 bytescount;
+
+	spcm = snd_sof_find_spcm_dai(scomp, rtd);
+	pcm_id = get_id_from_list(spcm->pcm.pcm_name);
+
+	bytescount = acp_get_byte_count(sdev, substream, pcm_id);
+	if (bytescount < 0)
+		return -EINVAL;
+	buffersize = frames_to_bytes(substream->runtime, substream->runtime->buffer_size);
+	pos = do_div(bytescount, buffersize);
+
+	return bytes_to_frames(substream->runtime, pos);
```

Comment:
I believe this is incorrect @ajitkupandey 

the .pointer in ALSA is the host-level DMA position, ie. the DMA that is used to feed data into the DSP memory.

the link level information is very useful, but it needs to be reported with the .delay field.
The difference between the host DMA position and link DMA position is the delay.

Reply:
see https://www.kernel.org/doc/html/latest/sound/designs/timestamping.html for a ascii-art that describes the different things you can report.

Reply:
Thanks..  we will take a look into this